### ISSUE TITLE
Deadlock during revert from the Version Browser

### ISSUE THREAD

**Issue by evanmiller** [jerrykrinock/BSManagedDocument#8]
I'm seeing a deadlock when attempting to revert to a previously autosaved version of a document.

Thread 0:

```
#0	0x00007fff618ae86a in __psynch_cvwait ()
#1	0x0000000100ad41b2 in _pthread_cond_wait ()
#2	0x00007fff37b4786f in -[__NSOperationInternal _waitUntilFinished:] ()
#3	0x00007fff339565d8 in -[_NSDocumentRevisionsLoadingOperation waitUntilFinished] ()
#4	0x00007fff3393cb7e in -[NSDocumentRevisionsController _requestUIForStackItem:wait:thenContinue:] ()
#5	0x00007fff3393cf9e in -[NSDocumentRevisionsController _loadSnapshotForStackItem:forceNew:wait:thenContinue:] ()
#6	0x00007fff3394a999 in -[NSDocumentRevisionsView ensureSnapshotForStackItem:forceNew:wait:] ()
#7	0x00007fff33950561 in -[NSDocumentRevisionsView _goToStackItemAtIndex:continuingPreviousAnimation:thenContinue:] ()
#8	0x00007fff33950cb1 in -[NSDocumentRevisionsView _animationToIndexCompleted:] ()
#9	0x00007fff3394a540 in __99-[NSDocumentRevisionsView _animateLayerAtIndex:fromPosition:toPosition:duration:completionHandler:]_block_invoke.1454 ()
#10	0x00007fff4030e4c9 in CA::Layer::run_animation_callbacks(void*) ()
#11	0x0000000100a56f1b in _dispatch_client_callout ()
#12	0x0000000100a659ee in _dispatch_main_queue_callback_4CF ()
#13	0x00007fff35842cd7 in __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ ()
#14	0x00007fff35842401 in __CFRunLoopRun ()
#15	0x00007fff358418be in CFRunLoopRunSpecific ()
#16	0x00007fff34b2d96b in RunCurrentEventLoopInMode ()
#17	0x00007fff34b2d6a5 in ReceiveNextEventCommon ()
#18	0x00007fff34b2d436 in _BlockUntilNextEventMatchingListInModeWithFilter ()
#19	0x00007fff32ec7987 in _DPSNextEvent ()
#20	0x00007fff32ec671f in -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] ()
#21	0x00007fff32ec083c in -[NSApplication run] ()
#22	0x00007fff32eafd7c in NSApplicationMain ()
#23	0x000000010009affa in main at /Users/emiller/Code/MagicMaps/main.m:27
#24	0x00007fff617763d5 in start ()
```

Thread 13:

```
#0	0x00007fff618ad9de in __ulock_wait ()
#1	0x0000000100a5780d in _dispatch_ulock_wait ()
#2	0x0000000100a578ec in _dispatch_thread_event_wait_slow ()
#3	0x0000000100a66c90 in __DISPATCH_WAIT_FOR_QUEUE__ ()
#4	0x0000000100a6646c in _dispatch_sync_f_slow ()
#5	0x00007fff353fa8b7 in -[NSManagedObjectContext performBlockAndWait:] ()
#6	0x00000001000d8525 in -[BSManagedDocument setManagedObjectContext:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:128
#7	0x00000001000d828a in -[BSManagedDocument managedObjectContext] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:113
#8	0x00000001000d8a0e in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:223
#9	0x0000000100021a61 in -[CKMapDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:2670
#10	0x00000001000d9073 in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:290
#11	0x00000001000d9ab4 in -[BSManagedDocument readFromURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:436
#12	0x000000010000dc36 in -[CKMapDocument readFromURL:ofType:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:438
#13	0x00007fff331ff169 in -[NSDocument _initWithContentsOfURL:ofType:error:] ()
#14	0x00007fff331ff04d in -[NSDocument initWithContentsOfURL:ofType:error:] ()
#15	0x000000010000b2d8 in -[CKMapDocument initWithContentsOfURL:ofType:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:151
#16	0x00007fff3328a2bd in -[NSDocumentController makeDocumentWithContentsOfURL:ofType:error:] ()
#17	0x00007fff339563f4 in -[_NSDocumentRevisionsLoadingOperation main] ()
#18	0x00007fff37ab51dd in -[__NSOperationInternal _start:] ()
#19	0x00007fff37adf197 in __NSOQSchedule_f ()
#20	0x0000000100a55e7c in _dispatch_call_block_and_release ()
#21	0x0000000100a56f1b in _dispatch_client_callout ()
#22	0x0000000100a5a2be in _dispatch_continuation_pop ()
#23	0x0000000100a593b4 in _dispatch_async_redirect_invoke ()
#24	0x0000000100a69824 in _dispatch_root_queue_drain ()
#25	0x0000000100a6a2da in _dispatch_worker_thread2 ()
#26	0x0000000100ad00b7 in _pthread_wqthread ()
#27	0x0000000100acfe01 in start_wqthread ()
```

Thread 16:

```
#0	0x00007fff618ad9de in __ulock_wait ()
#1	0x0000000100a5780d in _dispatch_ulock_wait ()
#2	0x0000000100a578ec in _dispatch_thread_event_wait_slow ()
#3	0x0000000100a66c90 in __DISPATCH_WAIT_FOR_QUEUE__ ()
#4	0x0000000100a6646c in _dispatch_sync_f_slow ()
#5	0x00007fff353fa8b7 in -[NSManagedObjectContext performBlockAndWait:] ()
#6	0x00000001000d8525 in -[BSManagedDocument setManagedObjectContext:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:128
#7	0x00000001000d828a in -[BSManagedDocument managedObjectContext] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:113
#8	0x00000001000d8a0e in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:223
#9	0x0000000100021a61 in -[CKMapDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:2670
#10	0x00000001000d9073 in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:290
#11	0x00000001000d9ab4 in -[BSManagedDocument readFromURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:436
#12	0x000000010000dc36 in -[CKMapDocument readFromURL:ofType:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:438
#13	0x00007fff331ff169 in -[NSDocument _initWithContentsOfURL:ofType:error:] ()
#14	0x00007fff331ff04d in -[NSDocument initWithContentsOfURL:ofType:error:] ()
#15	0x000000010000b2d8 in -[CKMapDocument initWithContentsOfURL:ofType:error:] at /Users/emiller/Code/MagicMaps/CKMapDocument.m:151
#16	0x00007fff3328a2bd in -[NSDocumentController makeDocumentWithContentsOfURL:ofType:error:] ()
#17	0x00007fff339563f4 in -[_NSDocumentRevisionsLoadingOperation main] ()
#18	0x00007fff37ab51dd in -[__NSOperationInternal _start:] ()
#19	0x00007fff37adf197 in __NSOQSchedule_f ()
#20	0x0000000100a55e7c in _dispatch_call_block_and_release ()
#21	0x0000000100a56f1b in _dispatch_client_callout ()
#22	0x0000000100a5a2be in _dispatch_continuation_pop ()
#23	0x0000000100a593b4 in _dispatch_async_redirect_invoke ()
#24	0x0000000100a69824 in _dispatch_root_queue_drain ()
#25	0x0000000100a6a2da in _dispatch_worker_thread2 ()
#26	0x0000000100ad00b7 in _pthread_wqthread ()
#27	0x0000000100acfe01 in start_wqthread ()
```

It appears that the main thread is coordinating two `NSDocumentRevisionsLoadingOperation`s, and waiting to them to finish, but they are waiting for the main queue to free up in order to execute `performBlockAndWait`.

I'm really not sure how to attack this - the `undoManager` needs to be accessed on the context's (main) queue. One possibility is to move the context to a private queue, but this may necessitate rewriting chunks of the application that assume the context and UI are executing on the same thread.

Another possibility is to disable the ignore `undoManager` when we're in Viewing mode, which is true of documents in the Version Browser.

**Comment by evanmiller** [comment]
I'm seeing a similar deadlock in the call to `setParentContext:` inside the same `setManagedObjectContext:` function, so skipping the undo manager creation will not necessarily solve the problem.

```
#0	0x00007fff618ad9de in __ulock_wait ()
#1	0x000000010113980d in _dispatch_ulock_wait ()
#2	0x00000001011398ec in _dispatch_thread_event_wait_slow ()
#3	0x0000000101148c90 in __DISPATCH_WAIT_FOR_QUEUE__ ()
#4	0x000000010114846c in _dispatch_sync_f_slow ()
#5	0x00007fff353fa8b7 in -[NSManagedObjectContext performBlockAndWait:] ()
#6	0x00007fff353fa6b2 in -[NSManagedObjectContext setParentContext:] ()
#7	0x00000001001e70c0 in -[BSManagedDocument setManagedObjectContext:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:167
#8	0x00000001001e6d3a in -[BSManagedDocument managedObjectContext] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:113
#9	0x00000001001e746e in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:227
#10	0x0000000100019481 in -[DSStatDocument configurePersistentStoreCoordinatorForURL:ofType:modelConfiguration:storeOptions:error:] at /Users/emiller/Code/Wizard/Datascope/DSStatDocument.m:521
#11	0x00000001001e7ad3 in -[BSManagedDocument configurePersistentStoreCoordinatorForURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:294
#12	0x00000001001e8404 in -[BSManagedDocument readFromURL:ofType:error:] at /Users/emiller/Code/BSManagedDocument.evanmiller/BSManagedDocument.m:440
#13	0x0000000100018795 in -[DSStatDocument readFromURL:ofType:error:] at /Users/emiller/Code/Wizard/Datascope/DSStatDocument.m:392
#14	0x00007fff331ff169 in -[NSDocument _initWithContentsOfURL:ofType:error:] ()
#15	0x00007fff331ff04d in -[NSDocument initWithContentsOfURL:ofType:error:] ()
#16	0x00007fff3328a2bd in -[NSDocumentController makeDocumentWithContentsOfURL:ofType:error:] ()
#17	0x00007fff339563f4 in -[_NSDocumentRevisionsLoadingOperation main] ()
#18	0x00007fff37ab51dd in -[__NSOperationInternal _start:] ()
#19	0x00007fff37adf197 in __NSOQSchedule_f ()
#20	0x0000000101137e7c in _dispatch_call_block_and_release ()
#21	0x0000000101138f1b in _dispatch_client_callout ()
#22	0x000000010113c2be in _dispatch_continuation_pop ()
#23	0x000000010113b45f in _dispatch_async_redirect_invoke ()
#24	0x000000010114b824 in _dispatch_root_queue_drain ()
#25	0x000000010114c2da in _dispatch_worker_thread2 ()
#26	0x00000001011b20b7 in _pthread_wqthread ()
#27	0x00000001011b1e01 in start_wqthread ()
```

**Comment by evanmiller** [comment]
I'm wondering if we can just have the document own the PSC; PSC's can be associated with multiple contexts, so we should be able to re-use the PSC when the context is manually set by the client.

The benefit of having a document-wide PSC is that `configurePersistentStoreCoordinatorForURL:...` will no longer need to call `.managedObjectContext`, which is the source of the deadlock when called on a non-main thread. We can also simplify the 10.6 compatibility code by using `dispatch_sync` to force main-thread execution on all platforms rather than testing for `performBlockAndWait:`.

**Comment by jerrykrinock** [comment]
Regarding the deadlock when reverting to a previously autosaved version:  Here is what I think you mean (Steps to Reproduce):

• Edit a document in your BSManagedDocument app.
• Switch to another app (which causes an autosave).
• Switch back to your BSManagedDocument app.
• Make another edit in the document.
• File > Browse to > Browse All Versions.
• Find and Restore the document which has only the first edit.

This does not deadlock in my app.  But my BSManagedDocument subclass may have some hacks which patches around the problem.  I remember waging some nasty battles years ago against Versions Browser :(  From your call stack, I see that you experience this in *your* app, not the BSManagedDocument demo app.  One of us needs to test this in the BSManagedDocument demo app.  Have I stated the Steps to Reproduce correctly?

**Comment by jerrykrinock** [comment]
@evanmiller: Regarding the similar deadlock in the call to `setParentContext:`, is this with the same Steps to Reproduce, that there are two deadocks in this situation?  Or does this occur in a different situation?

**Comment by jerrykrinock** [comment]
@evanmiller: Regarding *have the document own the PSC* and *PSC's can be associated with multiple contexts*, I don't understand that.

It looks like there are two managed object contexts per document – *parent* (`parentContext`) on a private queue, and the *main* (`context`) on the main queue.  The parent context gets its persistent store coordinator at line 161, `parentContext.persistentStoreCoordinator = coordinator`.  The main context gets its persistent store coordinator at line 171, `[context setPersistentStoreCoordinator:coordinator]`.  But the `coordinator` in each of these lines is the same, so I think we already have a *document-wide PSC*.

As I recall there is a new document instance for each revision being displayed, so having a *document-wide PSC* would not affect versions browsing either.

Help set me straight, please :)

**Comment by evanmiller** [comment]
> This does not deadlock in my app. But my BSManagedDocument subclass may have some hacks which patches around the problem. I remember waging some nasty battles years ago against Versions Browser :( From your call stack, I see that you experience this in _your_ app, not the BSManagedDocument demo app. One of us needs to test this in the BSManagedDocument demo app. Have I stated the Steps to Reproduce correctly?

Yes, though I did some manual saves in there too. The deadlock may arise when restoring a version while other versions are still loading from disk - in my app, documents can be large, and take a few seconds to load. You can see in the first set of backtraces that two separate threads (i.e. versions) are blocked on `setManagedObjectContext:`. So the code hasn't necessarily even attempted to load the requested version - it's still working through the queue of available versions to display and deadlocks because the main thread is no longer available.

**Comment by evanmiller** [comment]
> @evanmiller: Regarding the similar deadlock in the call to `setParentContext:`, is this with the same Steps to Reproduce, that there are two deadocks in this situation? Or does this occur in a different situation?

Yes, same steps to reproduce, just commented out the undoManager stuff to narrow the problem down.

**Comment by evanmiller** [comment]
> But the `coordinator` in each of these lines is the same, so I think we already have a _document-wide PSC_.

Well, a new coordinator is created every time the document's managedObjectContext is set. So in principle, the coordinator can change whenever someone manually sets the context.

In any event, I have a branch where the coordinator is an ivar created once for each document. I've been able to simplify some things and also circumvent the deadlock by avoiding calls to `[NSManagedObjectContext performBlockAndWait:]`.

Instead, in places, I use `[NSPersistentStoreCoordinator performBlockAndWait:]` - this was only introduced in 10.10, so I do harbor a concern that the following comment's demands may not be satisfied on earlier OS versions:

https://github.com/jerrykrinock/BSManagedDocument/blob/v0.3.x/BSManagedDocument.m#L225-L229

Issue #36 (referenced in the above comment) discusses both incarnations of `performBlockAndWait:`. 

I'm not sure about the proper path forward on 10.9 and earlier. If you'd like to discuss further, I can open a pull request with my changes and we can go from there.

**Comment by jerrykrinock** [comment]
I'm committed with a neighbor to do some tree trimming today, so it's wonderful to see you doing such nice work.  On first reading, your explanations all make sense.

For the record, I have no interest in going back any further than 10.10 which is my apps' current deployment target.  If it would make BSManagedDocument work better, I would be happy to it to up it to 10.12 or 10.13.

I should be able to test out your pull request in 24 hours.  My documents take several seconds to open too.  I shall create several versions and splice in a sleep() to stress test it.

**Comment by evanmiller** [comment]
Good to hear. My apps also target 10.10. I'd personally be more than happy to start a branch requiring 10.10, but I'd like to hear from others about their requirements before doing anything too drastic. The PR as it stands ought to work back to 10.6 (admittedly I haven't tested the changes except on 10.14).

**Comment by jerrykrinock** [comment]
Issue was fixed in commit 83fc8a6 which merged in evan-miller-document-owns-psc.
